packages/lib-webpack/src/webpack/DynamicRemotePlugin.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  extensionArraySchema,
+  remoteEntryScript,
+  remoteEntryCallback,
+} from '@openshift/dynamic-plugin-sdk';
+import type { Extension } from '@openshift/dynamic-plugin-sdk';
+import * as _ from 'lodash-es';
+import * as fs from 'fs';
+import * as glob from 'glob';
+import * as path from 'path';
+import * as webpack from 'webpack';
+import type { PluginBuildMetadata } from '../types/plugin';
+import { pluginBuildMetadataSchema } from '../yup-schemas';
+import { GenerateManifestPlugin } from './GenerateManifestPlugin';
+import { PatchContainerEntryPlugin } from './PatchContainerEntryPlugin';
+
+const parseJSONFile = (filePath: string) => JSON.parse(fs.readFileSync(filePath, 'utf-8'));
+
+const loadPluginMetadata = (fileName: string, baseDir = process.cwd()) => {
+  const filePath = path.resolve(baseDir, fileName);
+  return validatePluginMetadata(parseJSONFile(filePath));
+};
+
+const loadExtensions = (globPattern: string, baseDir = process.cwd()) => {
+  const filePaths = glob.sync(globPattern, { cwd: baseDir, absolute: true, nodir: true });
+  return _.flatMap(filePaths.map((filePath) => validateExtensions(parseJSONFile(filePath))));
+};
+
+const validatePluginMetadata = (obj: unknown): PluginBuildMetadata =>
+  pluginBuildMetadataSchema.strict(true).validateSync(obj);
+
+const validateExtensions = (obj: unknown): Extension[] =>
+  extensionArraySchema.strict(true).validateSync(obj);
+
+const getSharedModulesWithStrictSingletonConfig = (modules: string[]) =>
+  modules.reduce((acc, moduleRequest) => {
+    acc[moduleRequest] = {
+      // Enforce a single version of the shared module to be used at runtime
+      singleton: true,
+      // Prevent plugins from using a fallback version of the shared module
+      import: false,
+    };
+    return acc;
+  }, {} as WebpackSharedModules);
+
+/**
+ * Minimal subset of webpack `SharedConfig` type.
+ *
+ * @see https://webpack.js.org/plugins/module-federation-plugin/#sharing-hints
+ */
+type WebpackSharedModuleConfig = {
+  import?: string | false;
+  singleton?: boolean;
+};
+
+/**
+ * Equivalent to webpack `SharedObject` type.
+ */
+type WebpackSharedModules = { [moduleRequest: string]: WebpackSharedModuleConfig };
+
+export type DynamicRemotePluginOptions = Partial<{
+  /**
+   * Plugin metadata JSON file name, or the parsed plugin metadata.
+   *
+   * Default value: `plugin.json`.
+   */
+  pluginMetadata: string | PluginBuildMetadata;
```

Comment:
Based on @psav's suggestion, we can have two (optional but mutually exclusive) options:
```
pluginMetadata: PluginBuildMetadata;
pluginMetadataFile: string;
```

Reply:
In a different story / PR?

Reply:
Ticket: https://issues.redhat.com/browse/HAC-764